src/common/device_helpers.cuh

```diff
@@ -237,11 +237,19 @@ class MemoryLogger {
       peak_allocated_bytes =
         std::max(peak_allocated_bytes, currently_allocated_bytes);
       num_allocations++;
+      CHECK_GT(num_allocations, num_deallocations);
     }
-    void RegisterDeallocation(void *ptr) {
+    void RegisterDeallocation(void *ptr, size_t n, int current_device) {
+      auto itr = device_allocations.find(ptr);
+      if (itr == device_allocations.end()) {
+        LOG(FATAL) << "Attempting to deallocate " << n << " bytes on device "
```

Comment:
@sriramch See comments by @mt-jones in #4194 for possible explanation about deallocating memory on different device.  But adding these checks would make things clearer, maybe just not FATAL .

Reply:
i initially thought that the return addresses are *not* unique across devices, and hence added this check to be fatal, as we may be releasing memory that we did not intend to. perhaps, unified memory (with global addressing space) makes this moot. the intention here is to fail right at the place where we detect this and not let this fester beyond, as it provides better debuggability.